### PR TITLE
Add RPM testing framework for local development and CI preparation

### PR DESCRIPTION
**Pull request by darshdinger**
### Description
Adds comprehensive testing scripts for validating RPM package functionality locally. This provides the foundation for CI integration (planned separately) while enabling developers to test RPM changes before pushing.

### Changes
**Added:** `test/rpm/` directory with testing framework

### Test Scripts
- `quick_check.sh` - Fast validation of spec file requirements
- `test_docker_rpm.sh` - Docker-based isolated RPM lifecycle testing
- `test_manual_rpm.sh` - Interactive testing scenarios for edge cases
- `build_and_test.sh` - Orchestrates the full test suite
- `setup_test_environment.sh` - Automated prerequisite installation
- `README.md` - Complete testing documentation

### Testing Capabilities
- Validates systemd scriptlets are properly configured
- Verifies user/group requirements
- Tests package install/upgrade/removal scenarios
- Provides Docker-based isolation for clean testing
- Documents local testing procedures

### Usage
```bash
# Quick validation
./test/rpm/quick_check.sh

# Full test suite
./test/rpm/build_and_test.sh

# Interactive testing
./test/rpm/test_manual_rpm.sh
```

### References:
[EWM # 13112](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=13112)

### CODE REVIEW

File: test/rpm/setup_test_environment.sh [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+# Install required packages
+REQUIRED_PACKAGES=(
+    "rpm-build"
+    "rpmdevtools"
+    "systemd-rpm-macros"
+    "python3-devel"
+    "python3-setuptools"
+)
+
+echo -e "\n${YELLOW}Installing required packages...${NC}"
```
Comment (peterfpeterson) [inline test/rpm/setup_test_environment.sh]:
```
sudo dnf builddep ./livereduce.spec
```
Does the same thing without needing to keep this listing synchronized with the spec file

File: test/rpm/setup_test_environment.sh [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+NC='\033[0m'
+
+echo -e "${BLUE}===============================================${NC}"
+echo -e "${BLUE}    LiveReduce RPM Testing Environment Setup ${NC}"
+echo -e "${BLUE}===============================================${NC}"
+
+# Check if we're on a supported system
+if command -v dnf &>/dev/null; then
+    PKG_MANAGER="dnf"
+elif command -v yum &>/dev/null; then
```
Comment (peterfpeterson) [inline test/rpm/setup_test_environment.sh]: Drop support for yum. dnf is available since rhel8

File: test/rpm/quick_check.sh [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+set -e
+
+# Colors
+RED='\033[0;31m'
+GREEN='\033[0;32m'
+YELLOW='\033[1;33m'
+BLUE='\033[0;34m'
+NC='\033[0m'
+
+SPEC_FILE="/home/dzj/livereduce/livereduce.spec"
```
Comment (peterfpeterson) [inline test/rpm/quick_check.sh]: You hard-coded your home directory. A more dynamic way is (not tested)
```
SPECFILE="$(dirname $(dirname "$(realpath "$0")"))/livereduce.spec"
```

Comment (peterfpeterson) [reply test/rpm/quick_check.sh]: Since there are a couple of relative paths, maybe a variable defining the repository root would be useful

File: test/rpm/quick_check.sh [hunk 4]
```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
```
Comment (peterfpeterson) [inline test/rpm/quick_check.sh:1]: I think this would be more useful as a CI job. I like that it is easily accessible to developers for local debugging, but enforcing it through CI would make sure that it continues to work.

File: test/rpm/quick_check.sh (hunk 3 above)
Comment (darshdinger) [reply test/rpm/quick_check.sh]: Fixed this.

File: test/rpm/setup_test_environment.sh (hunk 2 above)
Comment (darshdinger) [reply test/rpm/setup_test_environment.sh]: Understood. Updated.

File: test/rpm/test_docker_rpm.sh [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    echo -e "\n${YELLOW}Cleaning up...${NC}"
+    docker stop "$CONTAINER_NAME" 2>/dev/null || true
+    docker rm "$CONTAINER_NAME" 2>/dev/null || true
+}
+
+# Set trap for cleanup
+trap cleanup EXIT
+
+# Build test environment
+echo -e "${YELLOW}Building Docker test environment...${NC}"
```
Comment (peterfpeterson) [inline test/rpm/test_docker_rpm.sh]: I don't like this method of embedding the Dockerfile in the middle of a script. It makes it harder to review. Please pull it out into a separate file.

File: test/rpm/test_docker_rpm.sh [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+cat > "$PROJECT_ROOT/Dockerfile.rpm-test" << 'EOF'
+FROM fedora:latest
+
+# Install required packages for RPM building and testing
+RUN dnf update -y && \
```
Comment (peterfpeterson) [inline test/rpm/test_docker_rpm.sh]: This should be split into things that are needed in the base image, and things that are required by the build. The latter should be installed via `dnf` from the buildrequires in the spec file.

File: test/rpm/test_docker_rpm.sh [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+EOF
+
+docker build -f "$PROJECT_ROOT/Dockerfile.rpm-test" -t "$DOCKER_IMAGE" "$PROJECT_ROOT"
+
+echo -e "${GREEN}✓ Docker image built successfully${NC}"
+
+# Run comprehensive tests
+echo -e "\n${YELLOW}Running comprehensive RPM tests...${NC}"
+
+docker run --name "$CONTAINER_NAME" \
```
Comment (peterfpeterson) [inline test/rpm/test_docker_rpm.sh]: This appears to be a lot of what is in #27

File: test/rpm/README.md [hunk 8]
```diff
@@ -0,0 +1,320 @@
+# RPM Testing Guide for LiveReduce
```
Comment (peterfpeterson) [inline test/rpm/README.md:1]: Please add a link to this document from the top level `README.md`. I believe you can use relative links and the web view will just work e.g.
```
[developer rpm testing readme](test/rpm/README.md)
```

File: test/rpm/setup_test_environment.sh [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+for group in "users" "hfiradmin"; do
+    if getent group "$group" &>/dev/null; then
+        echo -e "${GREEN}✓${NC} Group $group exists"
+    else
+        echo -e "${RED}✗${NC} Group $group not found"
+        all_installed=false
+    fi
+done
+
+# Create log directory with proper permissions
```
Comment (peterfpeterson) [inline test/rpm/setup_test_environment.sh]: From the top level README:
The logfile of what was setup for running, as well as other messages, is `/var/log/SNS_applications/livereduce.log` if run as the user `snsdata`, or `livereduce.log` in the current working directory (if run from the command line).

I'm not sure you need to do this bit of configuration. It is preferred if people don't routinely run as root for development. This reduces the chance of accidents that trash a whole system and increases the number of places it can be worked on.

File: test/rpm/test_docker_rpm.sh (hunk 5 above)
Comment (peterfpeterson) [reply test/rpm/test_docker_rpm.sh]: Please remove the embedded dockerfile

File: test/rpm/test_manual_rpm.sh [hunk 10]
```diff
@@ -0,0 +1,314 @@
+#!/bin/bash
```
Comment (peterfpeterson) [inline test/rpm/test_manual_rpm.sh]: I'm not sure we want to test the rpm this way. I like that it makes the rpm checks reproducible, but I don't know that we will be doing them repeatedly.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Some level of cleanup is appreciated before we go through another round of review.

The "exit code values and meanings" section of [this document](https://thelinuxcode.com/bash-script-returns-different-return-codes-exit/) suggest values to use. I don't know if we want to do a quick pass through the shell scripts to make our non-zero exits more correct.

File: test/rpm/README.md [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+#### RPM build fails
+- **Cause**: Missing build dependencies
+- **Solution**: `sudo dnf builddep -y ./livereduce.spec`
+
+### Debug Commands
+
+```bash
+# Check RPM dependencies
+rpm -qR python-livereduce
```
Comment (peterfpeterson) [inline test/rpm/README.md]: The end of `rpmbuild.sh` suggests using `rpm -qilRp` against the created rpm rather than having to install it. I think this should be
```suggestion
rpm -qRp python-livereduce*.rpm
```

File: test/rpm/README.md [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+# Verify installation
+rpm -V python-livereduce
```
Comment (peterfpeterson) [inline test/rpm/README.md]: This can also be run against the generated rpm
```suggestion
rpm -V python-livereduce*.rpm
```

File: test/rpm/README.md [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+# Check service status
+systemctl status livereduce.service
+
+# View service logs
+journalctl -u livereduce.service --no-pager
+
+# Validate service file
+systemd-analyze verify /usr/lib/systemd/system/livereduce.service
```
Comment (peterfpeterson) [inline test/rpm/README.md]: I just tried this on the version of the file in the repo (rather than installed) and it really hates our service file. Something to sort out later.

File: test/rpm/README.md [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+```
+
+## Development Notes
+
+- The RPM requires `nsd-app-wrap` which is SNS-specific and not available in public repositories
```
Comment (peterfpeterson) [inline test/rpm/README.md:192]: Is this the reason why we have to skip dependencies when installing in docker? The current rpm for that is ~800kB so we could, in principle, put it somewhere for test installs. This is a job for a later date.

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
Small changes to rpm debugging instructions

**Review by peterfpeterson — APPROVED** [review]
